### PR TITLE
Reorganize GitHub workflows.

### PR DESCRIPTION
**Pull request by Yury-Fridlyand**
Signed-off-by: Yury-Fridlyand <yuryf@bitquilltech.com>

## Description
* Remove obsolete workflows, mostly related to releasing ODFE artifacts.
* Update actions in rest workflows.
* Limit workflow scopes:
  * Reduce file scope, for example, JDBC workflow will be triggered on JDBC changes only.
  * Reduce action scope, so workflows shouldn't be triggered twice on PR.
* Strictly define env version to avoid unexpected failures.
* Upload test reports for SQL plugin workflow.
* Activate tests for SQL CLI.
* New workflow added which runs all tests for SQL plugin. Despite of the basic SQL workflow, it doesn't stop on the first failure and runs rests tests too. For more details see https://github.com/Bit-Quill/opensearch-project-sql/pull/119.
* New workflow would run on demand only. The button would appear after merge only.
image

See also discussion in [#118](https://github.com/Bit-Quill/opensearch-project-sql/pull/118) and [#119](https://github.com/Bit-Quill/opensearch-project-sql/pull/119).

## Issues Resolved
* JDBC/ODBC workflow won't be triggered on SQL plugin fixes and updates.
* Workflows shouldn't be triggered twice in PRs anymore.
* Cleanup.
* Activated CodeQL workflow.
* Activated SQL CLI testing.

## Not included changes
ODBC workflow still uses MacOS env v.10.15 which is obsolete. It fails on MacOS 11. We can't switch to newer MacOS version without changing/updating ODBC build process. This will be done in scope of another PR.
 
## Check List
- [x] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by MaxKsyunz — APPROVED** [review]

File: .github/workflows/sql-test-workflow.yml [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    - name: Set up JDK ${{ matrix.java }}
+      uses: actions/setup-java@v3
+      with:
+        distribution: 'temurin'
+        java-version: ${{ matrix.java }}
+    
+    - name: Run tests
+      id: tests
+      run: |
+        # checkstyle
```
Comment (MaxKsyunz) [inline .github/workflows/sql-test-workflow.yml:43]: Is this unpacking the build target? Just realized that whatever task that used to be, running it using `./gradlew --continue <task>` will keep running it after first failure.

Could you try that and see if it does what you are looking for? Sorry I didn't realize this earlier.

Comment (Yury-Fridlyand) [reply .github/workflows/sql-test-workflow.yml:43]: `--continue` says to ignore all failures and keep running.
Below is the output from `./gradlew --continue build`.
1. Too many garbage in the output. We have to parse it to provide a brief report.
2. Tasks executed in alphabetical order - `:core:x`, then `doctest`, `integ-test`, and so on, with `ppl` in the very end.

```
FAILURE: Build completed with 5 failures.

1: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':core:checkstyleMain'.
> Checkstyle rule violations were found. See the report at: file:///mnt/c/GitHub/opensearch-project-sql-bq/core/build/reports/checkstyle/main.html
  Checkstyle files with violations: 1
  Checkstyle violations by severity: [error:1]


* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
==============================================================================

2: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':core:checkstyleTest'.
> Checkstyle rule violations were found. See the report at: file:///mnt/c/GitHub/opensearch-project-sql-bq/core/build/reports/checkstyle/test.html
  Checkstyle files with violations: 1
  Checkstyle violations by severity: [error:1]


* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
==============================================================================

3: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':core:jacocoTestCoverageVerification'.
> Rule violated for class org.opensearch.sql.expression.datetime.DateTimeFunction: lines covered ratio is 0.9, but expected minimum is 1.0
  Rule violated for class org.opensearch.sql.expression.datetime.DateTimeFunction: branches covered ratio is 0.9, but expected minimum is 1.0

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
==============================================================================

4: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':doctest:startOpenSearch'.
> The process terminated unexpectedly - status code 1

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
==============================================================================

5: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':integ-test:integTest'.
> There were failing tests. See the report at: file:///mnt/c/GitHub/opensearch-project-sql-bq/integ-test/build/reports/tests/integTest/index.html

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
==============================================================================
```

File: .github/workflows/backport.yml [hunk 2]
```diff
@@ -7,7 +7,7 @@ on:
 
 jobs:
   backport:
-    runs-on: ubuntu-latest
+    runs-on: ubuntu-20.04
```
Comment (penghuo) [inline .github/workflows/backport.yml:10]: I am not sure about that, do we have any known issue regarding to ubuntu-latest. Any reason change to 20.04?
I think opensearch-project use ubuntu-latest by default.

File: .github/workflows/sql-cli-test-and-build-workflow.yml [hunk 3]
```diff
@@ -28,26 +37,21 @@ jobs:
...
-      #     sudo apt install openjdk-14-jdk
-      #     sudo apt install unzip
-      #     wget https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-oss-7.10.0-amd64.deb
-      #     sudo dpkg -i elasticsearch-oss-7.10.0-amd64.deb
-      #     sudo /usr/share/elasticsearch/bin/elasticsearch-plugin install --batch https://d3g5vo6xdbdb9a.cloudfront.net/downloads/elasticsearch-plugins/opensearch-sql/opensearch_sql-1.12.0.0.zip
-      #     sudo systemctl start elasticsearch.service
+        # tests are designed to run against http://localhost:9200, so we have to disable/remove security plugin
+      - name: Download and run OpenSearch
+        run: |
+          docker run -p 9200:9200 -e "discovery.type=single-node" -e "DISABLE_SECURITY_PLUGIN=true" --name test -d opensearchproject/opensearch:${{ matrix.opensearch-version }}
```
Comment (penghuo) [inline .github/workflows/sql-cli-test-and-build-workflow.yml:43]: it may not work. e.g, currently, SQL plugin depend on OpenSearch 2.4. But OpenSearch 2.4 Docker Images is not released yet.

File: .github/workflows/sql-test-and-build-workflow.yml [hunk 4]
```diff
@@ -32,13 +49,30 @@ jobs:
...
         token: ${{ secrets.CODECOV_TOKEN }}
 
     - name: Upload Artifacts
-      uses: actions/upload-artifact@v1
+      uses: actions/upload-artifact@v2
       with:
         name: opensearch-sql
         path: opensearch-sql-builds
+
+    - name: Upload test reports
```
Comment (penghuo) [inline .github/workflows/sql-test-and-build-workflow.yml:64]: why remove these?

Comment (penghuo) [inline .github/workflows/sql-test-and-build-workflow.yml:64]: why add this? upload to where?

Comment (Yury-Fridlyand) [reply .github/workflows/sql-test-and-build-workflow.yml:64]: It create a downloadable artifact, like you see there on bottom. Default lifetime of those is few days or one week.
https://github.com/Bit-Quill/opensearch-project-sql/actions/runs/3087022572

File: .github/workflows/backport.yml (hunk 2 above)
Comment (Yury-Fridlyand) [reply .github/workflows/backport.yml:10]: https://github.com/Bit-Quill/opensearch-project-sql/pull/118#discussion_r971385524
https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources
**TL;DR**
`ubuntu-latest` = `ubuntu-20.04` now, but subject to change

File: .github/workflows/sql-cli-test-and-build-workflow.yml (hunk 3 above)
Comment (Yury-Fridlyand) [reply .github/workflows/sql-cli-test-and-build-workflow.yml:43]: Currently, there are only `2.2.1` and `latest`. `latest` grabs the latest released version.
I can create another sub-job to download and build OpenSearch. So tests would run on the `current` version, e.g. 2.4 for now.

File: .github/workflows/draft-release-notes-workflow.yml [hunk 5]
```diff
@@ -1,14 +1,18 @@
...
 
 on:
+  pull_request:
+    types: [opened, reopened]
   push:
     branches:
-      - main
+      - 'main'
+      - '[1-9]+.[0-9x]+'
+      - '[1-9]+.[0-9x]+.[0-9x]+'
```
Comment (joshuali925) [inline .github/workflows/draft-release-notes-workflow.yml:10]: what is `'[1-9]+.[0-9x]+.[0-9x]+'` for?

File: .github/workflows/link-checker.yml [hunk 6]
```diff
@@ -1,14 +1,26 @@
...
+      - '[1-9]+.[0-9x]+'
+      - '[1-9]+.[0-9x]+.[0-9x]+'
+    paths:
+      - '**/*.pdf'
+      - '**/*.pnj'
+      - '**/*.jpg'
+      - '**/*.json'
+      - '**/*.md'
+      - '**/*.txt'
+      - '**/*.rst'
```
Comment (joshuali925) [inline .github/workflows/link-checker.yml:18]: if the workflow only triggers when we modify text, we cannot tell if any existing links became dead in other PRs?

also why `'**/*.jpg'` and what is .pnj?

File: .github/workflows/sql-cli-test-and-build-workflow.yml [hunk 7]
```diff
@@ -1,24 +1,33 @@
...
 
-    runs-on: ubuntu-latest
+    runs-on: ubuntu-20.04
     defaults:
       run:
         working-directory: sql-cli
     strategy:
       matrix:
         python-version: [3.8]
+        opensearch-version: [ 2.2.1, latest ]
```
Comment (joshuali925) [inline .github/workflows/sql-cli-test-and-build-workflow.yml:23]: why these two versions?

File: .github/workflows/sql-test-workflow.yml [hunk 8]
```diff
@@ -0,0 +1,104 @@
+name: SQL Plugin Tests
+
+on:
+  workflow_dispatch:
+  push:
+    branches:
+      - 'dev-*'
```
Comment (joshuali925) [inline .github/workflows/sql-test-workflow.yml:7]: i don't think we have `dev-` branch? and why 2 separate workflows for sql?

File: .github/workflows/sql-workbench-test-and-build-workflow.yml [hunk 9]
```diff
@@ -1,6 +1,21 @@
...
+    branches-ignore:
+      - 'dependabot/**'
+    paths:
+      - '**/*.java'
+      - '**/*.g4'
+      - '!sql-jdbc/**'
+      - '**gradle*'
+      - '**lombok*'
+      - '**/*.jar'
+      - '**/*.pom'
```
Comment (joshuali925) [inline .github/workflows/sql-workbench-test-and-build-workflow.yml:16]: i don't think these changes need to trigger workbench, since currently it only runs unit test not integration test?

**Review by joshuali925 — COMMENTED** [review]
a few questions

File: .github/workflows/draft-release-notes-workflow.yml (hunk 5 above)
Comment (Yury-Fridlyand) [reply .github/workflows/draft-release-notes-workflow.yml:10]: It is regex-like branch filter, it matches branch names like 3.3, 1.x, 2.3.x.

Comment (joshuali925) [reply .github/workflows/draft-release-notes-workflow.yml:10]: i meant to ask is this needed? since i think we only have x.y branch, not x.y.z

File: .github/workflows/link-checker.yml (hunk 6 above)
Comment (Yury-Fridlyand) [reply .github/workflows/link-checker.yml:18]: Do we have links that we need to check outside of text/docs? I'm ready to change the trigger criteria.
I included there pictures (`png`, `jpg`), because there is a number of documentation files which have pictures (links to picture files) ([sample](https://raw.githubusercontent.com/opensearch-project/sql/2.x/bi-connectors/TableauConnector/tableau_support.md)). This workflow checks these links as well.

`.pnj` is a mistake - fixed in 489f2adb. Thank you.

File: .github/workflows/sql-cli-test-and-build-workflow.yml (hunk 7 above)
Comment (Yury-Fridlyand) [reply .github/workflows/sql-cli-test-and-build-workflow.yml:23]: There is no particular reason. Should I place there all released versions?

**Review by MaxKsyunz — DISMISSED** [review]

File: .github/workflows/link-checker.yml (hunk 6 above)
Comment (joshuali925) [reply .github/workflows/link-checker.yml:18]: For example we have a link to `docs.opensearch.org` in some README.md, next day that link becomes `opensearch.org/docs` and `docs.opensearch.org` is no longer valid

But we will only find out our README.md is out of date when we modify a file in this list (which doesn't happen as frequent), so I'm saying should this run regardless of what changed?

File: .github/workflows/sql-cli-test-and-build-workflow.yml (hunk 7 above)
Comment (joshuali925) [reply .github/workflows/sql-cli-test-and-build-workflow.yml:23]: maybe just latest?

File: .github/workflows/sql-test-workflow.yml (hunk 8 above)
Comment (joshuali925) [reply .github/workflows/sql-test-workflow.yml:7]: I see, then is this needed (or should be renamed) in the upstream repo? we don't use `dev-*` branch, actually all develop and feature branches should contain a slash so we can remove them after they are merged

File: .github/workflows/link-checker.yml (hunk 6 above)
Comment (Yury-Fridlyand) [reply .github/workflows/link-checker.yml:18]: I understand. Agreed, I will change that.

File: .github/workflows/draft-release-notes-workflow.yml (hunk 5 above)
Comment (Yury-Fridlyand) [reply .github/workflows/draft-release-notes-workflow.yml]: Fixed in 63c51377.

File: .github/workflows/link-checker.yml (hunk 6 above)
Comment (Yury-Fridlyand) [reply .github/workflows/link-checker.yml]: Fixed in 63c51377.

File: .github/workflows/sql-test-workflow.yml (hunk 8 above)
Comment (Yury-Fridlyand) [reply .github/workflows/sql-test-workflow.yml]: Fixed in 63c51377. Workflow would run on demand only.

File: .github/workflows/sql-workbench-test-and-build-workflow.yml (hunk 9 above)
Comment (Yury-Fridlyand) [reply .github/workflows/sql-workbench-test-and-build-workflow.yml]: Thank you, fixed in 63c51377.

File: .github/workflows/sql-cli-test-and-build-workflow.yml (hunk 7 above)
Comment (Yury-Fridlyand) [reply .github/workflows/sql-cli-test-and-build-workflow.yml]: Fixed in 63c51377.

**Review by joshuali925 — APPROVED** [review]
thanks

**Review by penghuo — APPROVED** [review]

File: .github/workflows/sql-cli-test-and-build-workflow.yml (hunk 3 above)
Comment (acarbonetto) [reply .github/workflows/sql-cli-test-and-build-workflow.yml:42]: Might be worthwhile the next time someone has a chance. 
Another side-effect, is that we would have an artefact to download for our own local development usage... instead of having to download and setup the latest OpenSearch.

File: sql-cli/tests/test_opensearch_connection.py [hunk 10]
```diff
@@ -39,6 +39,7 @@ def test_query(self, connection):
         )
 
     @estest
+    @pytest.mark.skip(reason="Test is not compatible with OpenSearch >= 2.3.0, it returns HTTP/503 instead of HTTP/400")
```
Comment (acarbonetto) [inline sql-cli/tests/test_opensearch_connection.py:42]: I've left a comment on the bug to revert this if it's fixed: https://github.com/opensearch-project/sql/issues/851

**Review by acarbonetto — APPROVED** [review]
